middlewares/compress_test.go

```diff
@@ -0,0 +1,70 @@
+package middlewares
+
+import (
+	"bytes"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/NYTimes/gziphandler"
+	"github.com/containous/traefik/testhelpers"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	acceptEncodingHeader = "Accept-Encoding"
+	varyHeader           = "Vary"
+	gzip                 = "gzip"
+)
+
+func TestShouldCompressWhenNoContentEncodingHeader(t *testing.T) {
+	handler := &Compress{}
+
+	req := testhelpers.MustNewRequest(http.MethodGet, "http://localhost", nil)
+	req.Header.Add(acceptEncodingHeader, gzip)
+
+	baseBody := generateBytes(gziphandler.DefaultMinSize)
+	next := func(rw http.ResponseWriter, r *http.Request) {
+		rw.Write(baseBody)
+	}
+	rw := httptest.NewRecorder()
+
+	handler.ServeHTTP(rw, req, next)
+
+	assert.Equal(t, gzip, rw.Header().Get(contentEncodingHeader))
+	assert.Equal(t, acceptEncodingHeader, rw.Header().Get(varyHeader))
+
+	if bytes.Compare(rw.Body.Bytes(), baseBody) == 0 {
+		t.Errorf("got %v,\n expected a compressed body.", rw.Body.Bytes())
+	}
+}
+
+func TestShouldNotCompressWhenContentEncodingHeader(t *testing.T) {
+	handler := &Compress{}
+
+	req := testhelpers.MustNewRequest(http.MethodGet, "http://localhost", nil)
+	req.Header.Add(acceptEncodingHeader, gzip)
+	req.Header.Add(contentEncodingHeader, gzip)
+
+	baseBody := generateBytes(gziphandler.DefaultMinSize)
+
+	next := func(rw http.ResponseWriter, r *http.Request) {
+		rw.Write(baseBody)
+	}
+
+	rw := httptest.NewRecorder()
+	handler.ServeHTTP(rw, req, next)
+
+	assert.Equal(t, "", rw.Header().Get(contentEncodingHeader))
+	assert.Equal(t, "", rw.Header().Get(varyHeader))
+
+	assert.EqualValues(t, rw.Body.Bytes(), baseBody)
+}
```

Comment:
because it's not may way to test:
- I follow a Kent Beck rule about triangulation: I begin to thinking about a possible factorization if I repeat 3 times the same thing (the number 3, it's my limit, not a KB limit, but it's a very classic number for that with TDD)
- the tests are more flexible and more resistant to changes of the implementation.
- each test is independent and can be "recycle" (it's a testing strategy)
- the code in the test provide some limitation to be clearly factorize. (readability)
- table-driven test are not the only way to write good tests: each test case are different and we can use a lot of technique or approach.